dbms/src/Functions/FunctionsHashing.h

```diff
@@ -956,6 +956,136 @@ class FunctionURLHash : public IFunction
     }
 };
 
+struct JavaHashImpl
+{
+    static Int32 apply(const char * data, const size_t size)
+    {
+        Int32 h = 0;
+        for (int i = 0; i < (int)size; ++i)
+        {
+            h = 31 * h + data[i];
```

Comment:
signed integer overflow is undefined behaviour in C++

Reply:
You can accumulate in unsigned integer and then cast to signed.

Reply:
Using unsigned int and then cast, may get wrong result

Reply:
Using Int32 for accumulation with overflow is just illegal in C++ (in contrast to Java).